### PR TITLE
Foundations for non-linear solver and polymorphic application

### LINKED ISSUE #1317: Generic decorators don't work with generic functions
*UPDATE*: Although the original example is now special-cased in a plugin, the underlying problem is not solved, see https://github.com/python/mypy/issues/1317#issuecomment-341959279

The following:

``` python
from typing import *
from contextlib import *

T = TypeVar("T")

@contextmanager
def foo(x):
  # type: (T) -> Iterator[T]
  yield x

y = 1
with foo(2) as f:
  y = f
```

gives this error:

```
error: Incompatible types in assignment (expression has type "T", variable has type "int")
```

It should recognize the input as an `int` which would lead to the output being an `int` as well, not a `T`. This is probably due to the way the `contextmanager` stub is defined in typeshed.

### LINKED ISSUE #5738: Generic inference regression from #5699
Since #5699, mypy now rejects the following program (minimized from a few failures at dropbox):
```
from typing import TypeVar, Callable, Tuple

T = TypeVar('T')

def f(x: Callable[..., T]) -> T:
    return x()

x: Tuple[str, ...] = f(tuple)
```
with the error message
`Incompatible types in assignment (expression has type "Tuple[_T_co, ...]", variable has type "Tuple[str, ...]")`

@ilevkivskyi thoughts?

To be honest I'm not totally sure how we used to get this right, since my understanding was that we don't do the sort of instantiation of all type variables and unification that I would expect to be necessary?

### LINKED ISSUE #12919: Unexpected type binding when wrapping functions
**Bug Report**

I admit that I do not know if this is a bug, and not just my incomplete understanding of how the type system works, but I have run into some unexpected behaviour, and behaviour where mypy and pyre disagree, so it might be a bug.

I am writing code that lifts functions `T -> R` to `Optional[T] -> Optional[R]` in various ways, and when I write wrapper functions, type variables get bound in ways I didn't expect. It is particularly bad when I add protocols for various types on top of it, but I can create the unexpected behaviour without it. Basically, the issue is that if I write a function that changes a `Callable[[T], R]` into a `Callable[[Optional[T]], Optional[R]]`, and then apply such a function to another generic function `Callable[[T],R]` the type variables are no longer generic, and I cannot apply the wrapped function. This happens both when I work with wrapper functions or when I implement the wrapper as a `Generic` class.

**To Reproduce**

If I write a wrapper with a function, it might look like this:

```python
from typing import (
    TypeVar, 
    Protocol, Generic, 
    Optional as Opt
)


_T = TypeVar('_T')

class _F(Protocol[_T]):
    def __call__(self, __x: _T) -> _T:
        ...

# Lifting with a function...
def lift(f: _F[_T]) -> _F[Opt[_T]]:
    # This function would do more, but I just return f for the
    # type checking example
    return f # type: ignore
```

I get the desired behaviour if I wrap a function with a concrete type

```python
def f(x: int) -> int:
    return x

reveal_type(lift(f))  # _F[Opt[int]] -- as expected
```

but if I use a generic type, the wrapped function is not generic:

```python
def g(x: _T) -> _T:
    return x

reveal_type(lift(g)) # _F[Union[_T`-1,None]] -- _T is bound (incorrectly?)
                     # pyre thinks this is _F[None] which is also odd...
lift(g)(1)  # so won't accept int
```

**Expected Behavior**

I was expecting the wrapped function to have the type `Callable[[Opt[_T]], Opt[_T]]` and in particular I would expect to be able to call it with an `int` argument.

**To Reproduce**

If I use a `Generic` class for the lifted function, it can look like this:

```python
# Moving generic type to Generic...
class lifted(Generic[_T]):
    """Lift type _T to Opt[_T]."""

    def __init__(self, f: _F[_T]) -> None:
        self._f = f

    def __call__(self, x: Opt[_T], /) -> Opt[_T]:
        """Call lifted."""
        return self._f(x) # type: ignore

reveal_type(lifted[_T])  # _F[T?] -> lifted[_T?]
```

and it will work for a function with a concrete type, like `f: int -> int` above

```python
reveal_type(lifted(f))          # lifted[int] -- perfect
reveal_type(lifted(f).__call__) # Opt[int] -> Opt[int] -- perfect
lifted(f)(1)                    # We can call with int
lifted(f)(None)                 # We can call with None
```

but things get crazy when I use a generic function like `g: _T -> _T`:

```python
reveal_type(lifted(g))            # lifted[_T`1] -- _T is bound!
                                  # pyre thinks (correctly I think) it is lifted[_T]
reveal_type(lifted(g).__call__)   # Opt[_T`1] -> Opt[_T`1]
                                  # pyre says Any -> Any which I suppose is fine
lifted(g)(1)                      # incompatible type, int isn't an Opt[_T]
                                  # pyre seems to accep this one
lifted(g)(None)                   # but at least None is in Opt[_T]...

reveal_type(lifted[_T](g))          # lifted[_T?] -- Ok
reveal_type(lifted[_T](g).__call__) # def (None) ???
                                    # pyre: Opt[_T] -> Opt[_T] as expected

lifted[_T](g)(1)             # incompatible type "int"; expected "None"
lifted[_T](g)(None)          # but at least None works...
# Pyre doesn't like the [_T] here but accepts the calls without it...
```

**Expected Behavior**

Again, I would expect the `Generic` `lifted` class to accept types that match the wrapped function's type. This is more important when I need to add protocols to the types, but I would also expect it to work here.

**Your Environment**

I tested this in [mypy playground](https://mypy-play.net/?mypy=latest&python=3.10&gist=48dc95420afde2911df5b9a72f42e78c) and [pyre playground](https://pyre-check.org/play?input=from%20typing%20import%20(%0A%20%20%20%20TypeVar%2C%20%0A%20%20%20%20Protocol%2C%20Generic%2C%20%0A%20%20%20%20Optional%20as%20Opt%0A)%0A%0A%0A_T%20%3D%20TypeVar(%27_T%27)%0A%0Aclass%20_F(Protocol%5B_T%5D)%3A%0A%20%20%20%20def%20__call__(self%2C%20__x%3A%20_T)%20-%3E%20_T%3A%0A%20%20%20%20%20%20%20%20...%0A%0A%23%20Lifting%20with%20a%20function...%0Adef%20lift(f%3A%20_F%5B_T%5D)%20-%3E%20_F%5BOpt%5B_T%5D%5D%3A%0A%20%20%20%20%23%20This%20function%20would%20do%20more%2C%20but%20I%20just%20return%20f%20for%20the%0A%20%20%20%20%23%20type%20checking%20example%0A%20%20%20%20return%20f%20%23%20type%3A%20ignore%0A%0Adef%20f(x%3A%20int)%20-%3E%20int%3A%0A%20%20%20%20return%20x%0Adef%20g(x%3A%20_T)%20-%3E%20_T%3A%0A%20%20%20%20return%20x%0A%0Areveal_type(lift(f))%20%23%20_F%5BOpt%5Bint%5D%5D%20--%20as%20expected%0Areveal_type(lift(g))%20%23%20_F%5BUnion%5B_T%60-1%2CNone%5D%5D%20--%20_T%20is%20bound%20(incorrectly)%0Alift(g)(1)%20%20%20%20%20%20%20%20%20%20%20%23%20so%20won%27t%20accept%20int%0A%0A%0A%23%20Moving%20generic%20type%20to%20Generic...%0Aclass%20lifted(Generic%5B_T%5D)%3A%0A%20%20%20%20%22%22%22Lift%20type%20_T%20to%20Opt%5B_T%5D.%22%22%22%0A%0A%20%20%20%20def%20__init__(self%2C%20f%3A%20_F%5B_T%5D)%20-%3E%20None%3A%0A%20%20%20%20%20%20%20%20self._f%20%3D%20f%0A%0A%20%20%20%20def%20__call__(self%2C%20x%3A%20Opt%5B_T%5D%2C%20%2F)%20-%3E%20Opt%5B_T%5D%3A%0A%20%20%20%20%20%20%20%20%22%22%22Call%20lifted.%22%22%22%0A%20%20%20%20%20%20%20%20return%20self._f(x)%20%23%20type%3A%20ignore%0A%0Areveal_type(lifted%5B_T%5D)%20%20%23%20_F%5BT%3F%5D%20-%3E%20lifted%5B_T%3F%5D%0A%0Adef%20h(x%3A%20_T)%20-%3E%20_T%3A%0A%20%20%20%20return%20x%0A%0Areveal_type(lifted(f))%20%20%20%20%20%20%20%20%20%20%23%20lifted%5Bint%5D%20--%20perfect%0Areveal_type(lifted(f).__call__)%20%23%20Opt%5Bint%5D%20-%3E%20Opt%5Bint%5D%20--%20perfect%0Alifted(f)(1)%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20We%20can%20call%20with%20int%0Alifted(f)(None)%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20We%20can%20call%20with%20None%0A%0Areveal_type(lifted(g))%20%20%20%20%20%20%20%20%20%20%20%20%23%20lifted%5B_T%601%5D%20--%20_T%20is%20bound!%0Areveal_type(lifted(g).__call__)%20%20%20%23%20Opt%5B_T%601%5D%20-%3E%20Opt%5B_T%601%5D%0Alifted(g)(1)%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20incompatible%20type%2C%20int%20isn%27t%20an%20Opt%5B_T%5D%0Alifted(g)(None)%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20but%20at%20least%20None%20is%20in%20Opt%5B_T%5D...%0A%0Areveal_type(lifted%5B_T%5D(g))%20%20%20%20%20%20%20%20%20%20%23%20lifted%5B_T%3F%5D%20--%20Ok%0Areveal_type(lifted%5B_T%5D(g).__call__)%20%23%20def%20(None)%20%3F%3F%3F%0A%0Alifted(g)(1)%20%20%20%20%20%20%20%20%20%20%20%20%20%23%20incompatible%20type%20%22int%22%3B%20expected%20%22None%22%0Alifted(g)(None)%20%20%20%20%20%20%20%20%20%20%23%20but%20at%20least%20None%20works...%0A)

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #1317
Fixes #5738
Fixes #12919 
(also fixes a `FIX` comment that is more than 10 years old according to git blame)

Note: although this PR fixes most typical use-cases for type inference against generic functions, it is intentionally incomplete, and it is made in a way to limit implications to small scope.

This PR has essentially three components (better infer, better solve, better apply - all three are needed for this MVP to work):
* A "tiny" change to `constraints.py`: if the actual function is generic, we unify it with template before inferring constraints. This prevents leaking generic type variables of actual in the solutions (which makes no sense), but also introduces new kind of constraints `T <: F[S]`, where type variables we solve for appear in target type. These are much harder to solve, but also it is a great opportunity to play with them to prepare for single bin inference (if we will switch to it in some form later). Note unifying is not the best solution, but a good first approximation (see below on what is the best solution).
* New more sophisticated constraint solver in `solve.py`. The full algorithm is outlined in the docstring for `solve_non_linear()`. It looks like it should be able to solve arbitrary constraints that don't (indirectly) contain "F-bounded" things like `T <: list[T]`. Very short the idea is to compute transitive closure, then organize constraints by topologically sorted SCCs.
* Polymorphic type argument application in `checkexpr.py`. In cases where solver identifies there are free variables (e.g. we have just one constraint `S <: list[T]`, so `T` is free, and solution for `S` is `list[T]`) it will apply the solutions while creating new generic functions. For example, if we have a function `def [S, T] (fn: Callable[[S], T]) -> Callable[[S], T]` applied to a function `def [U] (x: U) -> U`, this will result in `def [T] (T) -> T` as the return.

I want to put here some thoughts on the last ingredient, since it may be mysterious, but now it seems to me it is actually a very well defined procedure. The key point here is thinking about generic functions as about infinite intersections or infinite overloads. Now reducing these infinite overloads/intersections to finite ones it is easy to understand what is actually going on. For example, imagine we live in a world with just two types `int` and `str`. Now we have two functions:
```python
T = TypeVar("T")
S = TypeVar("S")
U = TypeVar("U")

def dec(fn: Callable[[T], S]) -> Callable[[T], S]: ...
def id(x: U) -> U: ...
```
the first one can be seen as overload over
```
((int) -> int) -> ((int) -> int)  # 1
((int) -> str) -> ((int) -> str)  # 2
((str) -> int) -> ((str) -> int)  # 3
((str) -> str) -> ((str) -> str)  # 4
```
and second as an overload over
```
(int) -> int
(str) -> str
```
Now what happens when I apply `dec(id)`? We need to choose an overload that matches the argument (this is what we call type inference), but here is a trick, in this case two overloads of `dec` match the argument type. So (and btw I think we are missing this for real overloads) we construct a new overload that returns intersection of matching overloads `# 1` and `# 4`. So if we generalize this intuition to the general case, the inference is selection of an (infinite) parametrized subset among the bigger parameterized set of intersecting types. The only question is whether resulting infinite intersection is representable in our type system. For example `forall T. dict[T, T]` can make sense but is not representable, while `forall T. (T) -> T` is a well defined type. And finally, there is a very easy way to find whether a type is representable or not, we are already doing this during semantic analyzis. I use the same logic (that I used to view as ad-hoc because of lack of good syntax for callables) to bind type variables in the inferred type.

OK, so here is the list of missing features, and some comments on them:
1. Instead of unifying the actual with template we should include actual's variables in variable set we solve for, as explained in https://github.com/python/mypy/issues/5738#issuecomment-511242682. Note however, this will work only together with the next item
2. We need to (iteratively) infer secondary constraints after linear propagation, e.g. `Sequence[T] <: S <: Sequence[U] => T <: U`
3. Support `ParamSpec` (and probably `TypeVarTuple`). Current support for applying callables with `ParamSpec` to generics is hacky, and kind of dead-end. Although `(Callable[P, T]) -> Callable[P, List[T]]` works when applied to `id`, even a slight variation like `(Callable[P, List[T]]) -> Callable[P, T]` fails. I think it needs to be re-worked in the framework I propose (the tests I added are just to be sure I don't break existing code)
4. Support actual types that are generic in type variables with upper bounds or values (likely we just need to be careful when propagating constraints and choosing free variable within an SCC).
5. Add backtracking for upper/lower bound choice. In general, in the current "Hanoi Tower" inference scheme it is very hard to backtrack, but in in this specific choice in the new solver, it should be totally possible to switch from lower to upper bound on a previous step, if we found no solution (or `<nothing>`/`object`).
6. After we polish it, we can use the new solver in more situations, e.g. for return type context, and for unification during callable subtyping.
7. Long term we may want to allow instances to bind type variables, at least for things like `LRUCache[[x: T], T]`. Btw note that I apply force expansion to type aliases and callback protocols. Since I can't transform e.g. `A = Callable[[T], T]` into a generic callable without getting proper type.
8. We need to figure out a solution for scenarios where non-linear targets with free variables and constant targets mix without secondary constraints, like `T <: List[int], T <: List[S]`.

I am planning to address at least majority of the above items, but I think we should move slowly, since in my experience type inference is really fragile topic with hard to predict long reaching consequences. Please play with this PR if you want to and have time, and please suggest tests to add.

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
I will go through the `mypy_primer` either later today or tomorrow, it looks like there may be some regressions that need to be fixed.

**Comment by ilevkivskyi** [comment]
OK, so even before looking into `mypy_primer`, I found two bugs in my implementation:
* First, we may need multiple iterations to infer all variables within an SCC (simple example `T <: S, T <: int`)
* Second we may accidentally (depending on ordering within SCC) infer multiple free vars within an SCC, while there must always be at most one (simple example `T <: U, T <: S`)

These should be easy to fix.

Another thing is that my algorithm cannot solve a relatively simple system like `T <: Sequence[int], T <: Sequence[S]`, while there is an obvious solution `T = Sequence[int], S = int`. The problem is that we cannot "eagerly" solve using only constant targets, and then use these solutions to infer secondary constraints, this can lead to inferring invalid solutions. For example, in `T <: Sequence[int], T <: List[S], S <: bool`, we would wrongly infer `T = Sequence[int], S = bool`, while my algorithm finds the correct solution `T = List[bool], S = bool`. So we need to always "fold" the constraints starting from leaves, and only try some tricks when we have a mix of constant and non-linear targets with potentially free variables. This would be quite ugly, btw I found this on Wikipedia "Attempts to combine subtyping and type inference have caused quite some frustration", and now I understand it :-)

Btw I forgot to include one important missing feature in this MVP implementation: we need to support type variables with bounds/constraints. For example here:
```python
T = TypeVar("T")
S = TypeVar("S")
U = TypeVar("U", bound=int)

def dec(fn: Callable[[T], S]) -> Callable[[T], list[S]]: ...
def id(x: U) -> U: ...
```
the inferred type of `dec(id)` should also have the correct bound `def [T <: int] (T) -> list[T]`. Currently I think this will not work at all (actual will fail to unify I think), but even with a naive constraint propagation in the "proper" solution I mentioned, we would get unconstrained generic inferred, which would be wrong. I will edit the description to add this item so it will not be forgotten.

Finally, @JukkaL actually hiding this completely behind a flag will have one problem: this will require adding a flag to `infer_constraints()` that has ~50 recursive call sites. To clarify, it is not hard, it will just create a lot of "line noise".

**Comment by ilevkivskyi** [comment]
OK, so I spent some more time this weekend reading papers and thinking. Here are some comments on the three ingredients of this PR:
1. On the initial inference of constraints for generic actual: there is a quite simple way to implement proper constraint generation without full-scale secondary constraints, the key is that actual's variables cannot form chains, so I don't need to propagate. I will therefore add this to the PR.
2. On the new solver: it turns out the first step in my proposed algorithm (repeat propagation and secondary inference until there is no progress) is a well known thing and it is called *transitive closure* (and there is an efficient algorithm to generate it, but I will postpone it, since it is quite different from what I am doing now, and I anyway wanted to do proper secondary constraints later). Btw an interesting thing is that transitive closure is often considered as kind of a solution, for example in the problematic situation like `T :> List[int], T :> List[S]` people would say the solution is `S = <free>, T = Union[List[int], List[S]]`. This is however not acceptable for us since we don't support arbitrary universal quantification. But for the actual solution (what is usually called ground substitution) I didn't find anything apart from some heuristic. So I guess my way (topologically sorted SCCs) is fine.
3. On the polymorphic application: not really something useful, just an interesting thing is that GHC has a somewhat similar problem with "leaking" type variables: they use skolemization to support existential quantification, but then they need to be sure the variable used for this trick doesn't escape from the scope. They handle it with a thing called skolem levels check, which looks really like what I am proposing.

**Comment by ilevkivskyi** [comment]
> there is a quite simple way to implement proper constraint generation without full-scale secondary constraints, the key is that actual's variables cannot form chains, so I don't need to propagate. I will therefore add this to the PR.

I was being naive, it is not that trivial, so the plan stays the same (follow-up PRs will follow roughly in the order as they are listed in this PR's description, 1 and 2 will likely be combined into a single PR).

**Comment by ilevkivskyi** [comment]
OK, so I finally went through the `mypy_primer` output and here is the summary: Roughly half of the diff is good, i.e. either removed a false positive or new unused type ignore. The other half are new errors, all of which seem to be correct. They fall into two categories:

*First category* is self-types. Previously, for example the code like this worked
```python
from typing import Self, Callable, TypeVar

T = TypeVar("T")

def dec(f: Callable[..., T]) -> Callable[..., T]: ...

class C:
    @dec
    def test(self) -> Self: ...

c: C
reveal_type(c.test())  # Revealed type is "__main__.C"
```
Here is the old logic:
* Type of `test()` before decorator is `def [S] (S) -> S`
* When applying `dec` to `test` we used to infer `T = S` (leaking the type variable), and the type of `dec(test)` as `def (...) -> S` (a meaningless type in fact)
* Then due to how self-types are special-cased we would expand it to e.g. `def (...) -> C`.

Moreover, even more tricky decorators like
```python
T1 = TypeVar("T1")
T2 = TypeVar("T2")
def dec2(f: Callable[[T1], T2]) -> CustomProperty[T1, T2]: ...
```
used to (accidentally) work when applied to methods with self-types, while in fact we can't even express an inferred type of `dec2(test)` in current type system, it should be something like `forall S. CustomProperty[S, S]` (note btw item 7 in my master plan).

*Second category* is errors where a user writes a bad constructor signature in a generic class like this:
```python
class Value: ...
T = TypeVar("T", bound=Value)
class Box(Generic[T]):
    def __init__(self, item: Value) -> None: ...
```
Previously things like this worked:
```python
xs: List[Value]
items = map(Box, xs)
```
since we would infer (again, a meaningless) type `map[Box[T]]` for `items` (which btw may even not cause many errors, because of the upper bound). Now we correctly infer `map[Box[<nothing>]]` and therefore ask for an annotation.

For the first issue I added a special case that should cover the "new style" self-types (by simply excluding them from new inference logic), just to minimize the fallout. For the second issue, I don't think we need to do anything, the old behavior is clearly wrong.

Soon I am going to hide the new inference logic behind an internal flag (btw I think I will use `typestate` instead of piping the flag through 50+ function calls, since it is temporary anyway).

**Comment by JukkaL** [comment]
Apologies for the delay, I was supposed to review this today, but I didn't have enough uninterrupted time. I've reserved a few hours tomorrow to review this PR. Thanks for your patience.

File: mypy/solve.py [hunk 1]
```diff
@@ -33,62 +41,332 @@ def solve_constraints(
...
+    elif is_subtype(bottom, top):
+        candidate = bottom
+    else:
+        candidate = None
+    return candidate
+
+
+def normalize_constraints(
+    constraints: list[Constraint], vars: list[TypeVarId]
+) -> list[Constraint]:
```
Comment (JukkaL) [inline mypy/solve.py:242]: It would nice to have some unit tests for this (not need to do this in this PR).

File: mypy/solve.py [hunk 2]
```diff
@@ -33,62 +41,332 @@ def solve_constraints(
...
+    res = constraints.copy()
+    for c in constraints:
+        if isinstance(c.target, TypeVarType):
+            res.append(Constraint(c.target, neg_op(c.op), c.origin_type_var))
+    return [c for c in remove_dups(constraints) if c.type_var in vars]
+
+
+def propagate_constraints_for(
+    var: TypeVarId, direction: int, cmap: dict[TypeVarId, list[Constraint]]
+) -> list[Constraint]:
```
Comment (JukkaL) [inline mypy/solve.py:258]: Similarly, it would be nice to eventually have some unit tests for this. The logic is a little tricky to follow.

File: mypy/solve.py [hunk 3]
```diff
@@ -33,62 +41,332 @@ def solve_constraints(
...
+    return extra_constraints
+
+
+def transitive_closure(
+    tvars: list[TypeVarId], constraints: list[Constraint]
+) -> tuple[dict[TypeVarId, set[Type]], dict[TypeVarId, set[Type]]]:
+    """Find transitive closure for given constraints on type variables.
+
+    Transitive closure gives maximal set of lower/upper bounds for each type variable, such
+    we cannot deduce any further bounds by chaining other existing bounds.
```
Comment (JukkaL) [inline mypy/solve.py:297]: Give an example or two that illustrates what this means in the docstring? Again, it would be good to have some unit tests to validate the logic and avoid regressions in the future.

File: mypy/solve.py [hunk 4]
```diff
@@ -33,62 +41,332 @@ def solve_constraints(
...
+            cmap[con.type_var].append(con)
+
+    dmap = compute_dependencies(cmap)
+    sccs = list(strongly_connected_components(set(vars), dmap))
+    if all(check_linear(scc, cmap) for scc in sccs):
+        raw_batches = list(topsort(prepare_sccs(sccs, dmap)))
+        leafs = raw_batches[0]
+        free_vars = []
+        for scc in leafs:
+            if all(
```
Comment (JukkaL) [inline mypy/solve.py:115]: Add comment here about the purpose of this if statement (I think I figured it out but it wasn't obvious).

File: mypy/solve.py [hunk 5]
```diff
@@ -33,62 +41,332 @@ def solve_constraints(
...
+        uppers = {tv: {expand_type(u, subs) for u in uppers[tv]} for tv in uppers}
+        for v in cmap:
+            for c in cmap[v]:
+                c.target = expand_type(c.target, subs)
+    return solutions
+
+
+def solve_one(
+    lowers: Iterable[Type], uppers: Iterable[Type], not_allowed_vars: list[TypeVarId]
+) -> Type | None:
```
Comment (JukkaL) [inline mypy/solve.py:186]: Similarly some unit tests would be nice at some point.

File: mypy/solve.py [hunk 6]
```diff
@@ -33,62 +41,332 @@ def solve_constraints(
...
+                    lowers[u] |= lowers[lower]
+            for l in tvars:
+                if (l, lower) in graph:
+                    uppers[l] |= uppers[upper]
+    return lowers, uppers
 
+
+def compute_dependencies(
+    cmap: dict[TypeVarId, list[Constraint]]
+) -> dict[TypeVarId, list[TypeVarId]]:
```
Comment (JukkaL) [inline mypy/solve.py:340]: I'm repeating myself, but again unit tests would be nice :-)

File: test-data/unit/check-generics.test [hunk 7]
```diff
@@ -2733,3 +2733,229 @@ dict1: Any
...
+
+S = TypeVar('S')
+T = TypeVar('T')
+U = TypeVar('U')
+
+def dec(f: Callable[[S], T]) -> Callable[[S], List[T]]:
+    ...
+def id(x: U) -> U:
+    ...
+reveal_type(dec(id))  # N: Revealed type is "def [S] (S`1) -> builtins.list[S`1]"
```
Comment (JukkaL) [inline test-data/unit/check-generics.test:2796]: Can you add an actual use of the decorator to at least some of the decorator-like use cases, for more end-to-end testing? E.g. something like this:
```py
@dec 
def f(...) -> ...: ...  # Generic function
reveal_type(f(...))
```

**Review by JukkaL — APPROVED** [review]
Looks great! It's exciting that we will be able to support more complex decorator use cases. Also it's great to have type variable leakage bugs fixed, as they were annoying and super confusing. This opens up many additional use cases. Also I like that you are working on improvements incrementally and the new feature is behind a feature flag until we have more confidence.

This was a very complex piece work. Most of my comments are about adding comments or unit tests to make the code easier to understand and maintain in the long term. Feel free to merge this PR without making all the suggested changes, since the feature is not enabled by default.

Thank you for working on this!

File: mypy/solve.py (hunk 1 above)
Comment (JukkaL) [reply mypy/solve.py:242]: On reflection, I'm not sure if it's worth having unit tests for all of these functions. It may make more sense to have some unit tests for top-level functions. @jhance has recently written some unit tests that target type inference in `mypy/test/testconstraints.py`, and I think that they are helpful, even if we are not trying to cover all edge cases.

Comment (ilevkivskyi) [reply mypy/solve.py:242]: FWIW I think it is a good idea. Of course not _all_ of these really needs to be tested, but for most of them it is a good thing. Especially for e.g. `transitive_closure()` since logic there is really easy to get wrong. I will however add them in a separate PR.

**Comment by ilevkivskyi** [comment]
@JukkaL thanks for review! I will address everything except the unit tests soon. Unit tests will go into a separate PR.
